libs/libc/tls/tls_setdtor.c

```diff
@@ -59,18 +56,16 @@
  *
  ****************************************************************************/
 
-int tls_set_dtor(int tlsindex, tls_dtor_t destr)
+int tls_set_dtor(int tlsindex, tls_dtor_t dtor)
 {
-  FAR struct tcb_s *rtcb = this_task();
-  FAR struct task_group_s *group = rtcb->group;
-  irqstate_t flags;
+  FAR struct task_info_s *tinfo = task_get_info();
 
-  DEBUGASSERT(group != NULL);
+  DEBUGASSERT(tinfo != NULL);
   DEBUGASSERT(tlsindex >= 0 && tlsindex < CONFIG_TLS_NELEM);
 
-  flags = spin_lock_irqsave(NULL);
-  group->tg_tlsdestr[tlsindex] = destr;
-  spin_unlock_irqrestore(NULL, flags);
+  sem_wait(&tinfo->ta_tlssem);
+  tinfo->ta_tlsdtor[tlsindex] = dtor;
+  sem_post(&tinfo->ta_tlssem);
```

Comment:
You don't need to take the semaphore here.  The assignment is an atomic operation.